py/server/deephaven/barrage.py

```diff
@@ -0,0 +1,222 @@
+_JTableSpec = jpy.get_type("io.deephaven.qst.table.TableSpec")
+_JBarrageTableResolver = jpy.get_type("io.deephaven.server.uri.BarrageTableResolver")
+_JChannelHelper = jpy.get_type("io.deephaven.client.impl.ChannelHelper")
+_JDeephavenChannelImpl = jpy.get_type("io.deephaven.proto.DeephavenChannelImpl")
+_JSessionImpl = jpy.get_type("io.deephaven.client.impl.SessionImpl")
+_JExecutors = jpy.get_type("java.util.concurrent.Executors")
+_JRootAllocator = jpy.get_type("org.apache.arrow.memory.RootAllocator")
+_JDeephavenApiServer = jpy.get_type("io.deephaven.server.runner.DeephavenApiServer")
+
+
+class BarrageSession():
+    """ A Deephaven Barrage session to a remote server."""
+
+    def __init__(self, j_barrage_session: jpy.JType, j_managed_channel: jpy.JType = None):
+        """ Initializes a Barrage session.
+
+        when BarrageSession is created via the barrage_session() factory function, j_managed_channel is always provided,
+        and when BarrageSession.close() is called, it will shut down the channel as well as close the j_barrage_session..
+
+        when BarrageSession is initialized directly and j_managed_channel is None, when BarrageSession.close() is called,
+        it will only close the j_barrage_session, it is the responsibility of the user to shut down the associated
+        channel if needed.
+        """
+        self.j_barrage_session = j_barrage_session
+        self.j_session = j_barrage_session.session()
+        self.j_managed_channel = j_managed_channel
+
+    def __enter__(self) -> BarrageSession:
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        self.close()
+
+    def close(self) -> None:
+        """ Closes the barrage session.
+
+        If the BarrageSession is initialized with a managed channel, the channel will be shut down as well.
+        """
+        try:
+            self.j_barrage_session.close()
+            if self.j_managed_channel:
+                self.j_managed_channel.shutdownNow()
+                self.j_managed_channel.awaitTermination(10, _JTimeUnit.SECONDS)
+        except Exception as e:
+            raise DHError(e, "failed to close the barrage session.") from e
+
+    def subscribe(self, ticket: bytes) -> Table:
+        """ Subscribes to a published remote table with the given ticket.
+
+        Note, if the remote table is closed or its owner session is closed, the ticket becomes invalid. If the same
+        ticket is subscribed to multiple times, multiple subscriptions will be created.
+
+        Args:
+            ticket (bytes): the bytes of the ticket
+
+        Returns:
+            a Table
+
+        Raises:
+            DHError
+        """
+        try:
+            j_table_handle = self.j_session.of(_JTableSpec.ticket(ticket))
+            j_barrage_subscription = self.j_barrage_session.subscribe(j_table_handle,
+                                                                      _JBarrageTableResolver.SUB_OPTIONS)
```

Comment:
Sorry I didn't notice this earlier; we'll be leaking exports this way. We should be calling subscribe w/ the TableSpec directly.